atlasdb-client/src/test/java/com/palantir/atlasdb/util/TopNMetricPublicationControllerTest.java

```diff
@@ -0,0 +1,153 @@
+                IntStream.rangeClosed(1, 50).boxed(), Comparator.naturalOrder(), 0))
+                .isInstanceOf(IllegalStateException.class)
+                .hasMessageContaining("The order statistic to be queried for must be positive");
+        assertThatThrownBy(() -> TopNMetricPublicationController.calculateOrderStatistic(
+                IntStream.rangeClosed(1, 50).boxed(), Comparator.naturalOrder(), -1))
+                .isInstanceOf(IllegalStateException.class)
+                .hasMessageContaining("The order statistic to be queried for must be positive");
+    }
+
+    @Test
+    public void canSelectivelyPublishGaugeResults() {
+        AtomicLong value1 = new AtomicLong(42);
+        AtomicLong value2 = new AtomicLong(157);
+
+        Gauge<Long> gauge1 = value1::get;
+        Gauge<Long> gauge2 = value2::get;
+
+        TopNMetricPublicationController<Long> controller = TopNMetricPublicationController.create(1);
+        MetricPublicationFilter filter1 = controller.registerAndCreateFilter(gauge1);
+        MetricPublicationFilter filter2 = controller.registerAndCreateFilter(gauge2);
+
+        assertThat(filter1.shouldPublish()).isFalse();
+        assertThat(filter2.shouldPublish()).isTrue();
+    }
+
+    @Test
+    public void canSelectivelyPublishMultipleGaugeResults() {
+        List<Gauge<Long>> atomicLongs = LongStream.range(0, 100)
+                .mapToObj(AtomicLong::new)
+                .<Gauge<Long>>map(atomicLong -> atomicLong::get)
+                .collect(Collectors.toList());
+
+        TopNMetricPublicationController<Long> controller = TopNMetricPublicationController.create(7);
+
+        List<MetricPublicationFilter> filtersInOrder = atomicLongs.stream()
+                .map(controller::registerAndCreateFilter)
+                .collect(Collectors.toList());
+
+        assertThat(filtersInOrder.get(0).shouldPublish()).isFalse();
+        assertThat(filtersInOrder.get(50).shouldPublish()).isFalse();
+        assertThat(filtersInOrder.get(92).shouldPublish()).isFalse();
+        assertThat(filtersInOrder.get(93).shouldPublish()).isTrue();
+        assertThat(filtersInOrder.get(94).shouldPublish()).isTrue();
+        assertThat(filtersInOrder.get(99).shouldPublish()).isTrue();
+    }
+
+    @Test
+    public void publishesAllGaugesIfThereAreFewerThanTheThreshold() {
+        AtomicLong value1 = new AtomicLong(42);
+        AtomicLong value2 = new AtomicLong(157);
+
+        Gauge<Long> gauge1 = value1::get;
+        Gauge<Long> gauge2 = value2::get;
+
+        TopNMetricPublicationController<Long> controller = TopNMetricPublicationController.create(50);
+        MetricPublicationFilter filter1 = controller.registerAndCreateFilter(gauge1);
+        MetricPublicationFilter filter2 = controller.registerAndCreateFilter(gauge2);
+
+        assertThat(filter1.shouldPublish()).isTrue();
+        assertThat(filter2.shouldPublish()).isTrue();
+    }
+
```

Comment:
can we also have a test that checks lower bound is refreshed?

Reply:
Will do!